### PR TITLE
[WIP] Asset-buyer Refactor for CFL

### PR DESCRIPTION
**Pull request by dave4506**
## Description

Still have yet to test all the code, but the large changes made are as follows:
- Changed naming schemes from `AssetBuyer` to `SwapQuoter`
- Changed `BuyQuote` to `SwapQuote`
- Added functionality to support ERC to ERC pairs for requesting `SwapQuotes`
- Created a new interface `SwapQuoteConsumer` that consumes or executes a `SwapQuote`
- Created preliminary Forwarder `SwapQuoteConsumer` that consumes valid forwarder `SwapQuotes` and returns calldata, smart contract params, or executes a forwarder smart contract call. 

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-buyer/src/constants.ts [hunk 1]
```diff
@@ -1,40 +1,35 @@
...
+import { AssetSwapQuoterOpts, ForwarderSwapQuoteExecutionOpts, OrdersAndFillableAmounts, SwapQuoteRequestOpts, SwapQuoteExecutionOpts } from './types';
 
 const NULL_ADDRESS = '0x0000000000000000000000000000000000000000';
 const MAINNET_NETWORK_ID = 1;
 
-const DEFAULT_ASSET_BUYER_OPTS: AssetBuyerOpts = {
+const DEFAULT_ASSET_SWAP_QUOTER_OPTS: AssetSwapQuoterOpts = {
     networkId: MAINNET_NETWORK_ID,
     orderRefreshIntervalMs: 10000, // 10 seconds
     expiryBufferSeconds: 120, // 2 minutes
```
Comment (fabioberger) [inline packages/asset-buyer/src/constants.ts]: Should we standardize to a single unit?

File: packages/asset-buyer/src/constants.ts [hunk 2]
```diff
@@ -1,40 +1,35 @@
...
 import { BigNumber } from '@0x/utils';
 
-import { AssetBuyerOpts, BuyQuoteExecutionOpts, BuyQuoteRequestOpts, OrdersAndFillableAmounts } from './types';
+import { AssetSwapQuoterOpts, ForwarderSwapQuoteExecutionOpts, OrdersAndFillableAmounts, SwapQuoteRequestOpts, SwapQuoteExecutionOpts } from './types';
 
 const NULL_ADDRESS = '0x0000000000000000000000000000000000000000';
 const MAINNET_NETWORK_ID = 1;
 
-const DEFAULT_ASSET_BUYER_OPTS: AssetBuyerOpts = {
+const DEFAULT_ASSET_SWAP_QUOTER_OPTS: AssetSwapQuoterOpts = {
```
Comment (fabioberger) [inline packages/asset-buyer/src/constants.ts]: If these are passed into the methods that are part of the exported interface, these types should also be exported. Additionally, each of the fields should have a corresponding comment above this declaration (these will show up in the docs page).

File: packages/asset-buyer/src/order_providers/basic_order_provider.ts [hunk 3]
```diff
@@ -38,4 +38,13 @@ export class BasicOrderProvider implements OrderProvider {
         const ordersWithTakerAssetData = _.filter(this.orders, { takerAssetData });
         return _.map(ordersWithTakerAssetData, order => order.makerAssetData);
     }
+    /**
+     * Given a maker asset data string, return all availabled paired taker asset data strings.
+     * @param   makerAssetData   A string representing the maker asset data.
+     * @return  An array of asset data strings that can be used to purchased makerAssetData.
+     */
+    public async getAvailableTakerAssetDatasAsync(makerAssetData: string): Promise<string[]> {
```
Comment (fabioberger) [inline packages/asset-buyer/src/order_providers/basic_order_provider.ts]: What about multiAsset proxy?

File: packages/asset-buyer/src/order_providers/standard_relayer_api_order_provider.ts [hunk 4]
```diff
@@ -111,4 +111,32 @@ export class StandardRelayerAPIOrderProvider implements OrderProvider {
             }
         });
     }
+    /**
+     * Given a maker asset data string, return all availabled paired taker asset data strings.
```
Comment (fabioberger) [inline packages/asset-buyer/src/order_providers/standard_relayer_api_order_provider.ts]:
```suggestion
     * Given a maker asset data string, return all available paired taker asset data strings.
```

File: packages/asset-buyer/src/order_providers/standard_relayer_api_order_provider.ts [hunk 5]
```diff
@@ -111,4 +111,32 @@ export class StandardRelayerAPIOrderProvider implements OrderProvider {
...
+     * @param   makerAssetData   A string representing the maker asset data.
+     * @return  An array of asset data strings that can be used to purchased makerAssetData.
+     */
+    public async getAvailableTakerAssetDatasAsync(makerAssetData: string): Promise<string[]> {
+        // Return a maximum of 1000 asset datas
+        const maxPerPage = 1000;
```
Comment (fabioberger) [inline packages/asset-buyer/src/order_providers/standard_relayer_api_order_provider.ts]: This is hidden from the caller and not immediately obvious. Let's make it a optional param that details to 1000.

File: packages/asset-buyer/src/order_providers/standard_relayer_api_order_provider.ts [hunk 6]
```diff
@@ -111,4 +111,32 @@ export class StandardRelayerAPIOrderProvider implements OrderProvider {
...
+        const assetPairsRequest = { assetDataA: makerAssetData };
+        const fullRequest = {
+            ...requestOpts,
+            ...assetPairsRequest,
+        };
+        let response: AssetPairsResponse;
+        try {
+            response = await this._sraClient.getAssetPairsAsync(fullRequest);
+        } catch (err) {
+            throw new Error(AssetSwapQuoterError.StandardRelayerApiError);
```
Comment (fabioberger) [inline packages/asset-buyer/src/order_providers/standard_relayer_api_order_provider.ts]: Are we sure we want to just swallow the actual error returned? Might make it harder to debug the actual issue. Perhaps we log it?

File: packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    SwapQuoteExecutionOpts,
+    SwapQuoteGetOutputOpts,
+    SwapQuoteInfo,
+} from '../types';
+import { assert } from '../utils/assert';
+import { utils } from '../utils/utils';
+
+import { assetDataUtils } from '../utils/asset_data_utils';
+
+export interface ForwarderSwapQuoteGetOutputOpts extends SwapQuoteGetOutputOpts {
```
Comment (fabioberger) [inline packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts]: Add comments explaining each opt.

File: packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    feePercentage: number;
+    feeRecipient: string;
+    ethAmount?: BigNumber;
+}
+
+export interface ForwarderSwapQuoteExecutionOpts extends SwapQuoteExecutionOpts, ForwarderSwapQuoteGetOutputOpts {
+}
```
Comment (fabioberger) [inline packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts]: Do you need the empty braces?

File: packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+
+const addAffiliateFeeToSwapQuoteInfo = (quoteInfo: SwapQuoteInfo, feePercentage: number): SwapQuoteInfo => {
```
Comment (fabioberger) [inline packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts]: Why are you using anonymous functions here?

File: packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        supportedProvider: SupportedProvider,
+        options: Partial<SwapQuoteConsumerOpts> = {},
+    ) {
+        const { networkId } = _.merge(
+            {},
+            constants.DEFAULT_ASSET_SWAP_QUOTER_OPTS,
+            options,
+        );
+        const provider = providerUtils.standardizeOrThrow(supportedProvider);
+        assert.isNumber('networkId', networkId);
```
Comment (fabioberger) [inline packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts]: Nit: Let's put this up one line. Try to group logic as closely together as possible.

File: packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+    }
+
+    public getCalldataOrThrow = (quote: SwapQuote, opts: Partial<ForwarderSwapQuoteGetOutputOpts>): CalldataInformation => {
+        if (!this.isValidForwarderSwapQuote(quote)) {
+            throw new Error(AssetSwapQuoterError.InvalidForwarderSwapQuote);
+        }
+        const { params, to, value } = this.getSmartContractParamsOrThrow(quote, opts);
+        const methodAbi = utils.getMethodAbiFromContractAbi(this._contractWrappers.forwarder.abi, 'marketBuyOrdersWithEth') as MethodAbi;
+        const abiEncoder = new AbiEncoder.Method(methodAbi);
+        const args = [params.orders, params.makerAssetFillAmount, params.signatures, params.feeOrders, params.feeSignatures, params.feePercentage, params.feeRecipient];
```
Comment (fabioberger) [inline packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts]: Are you using the Prettier extension? This line is way too long.

File: packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+        const calldataHexString = abiEncoder.encode(args);
+        return {
+            calldataHexString,
+            to,
+            value,
+        };
+    }
+
```
Comment (fabioberger) [inline packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts]: Nit: Remove these new lines. I believe this is an [unenforced coding convention](https://github.com/0xProject/0x-monorepo/blob/development/CONTRIBUTING.md#unenforced-coding-conventions). You should also set up the [auto-formatter](https://github.com/0xProject/0x-monorepo/blob/development/CONTRIBUTING.md#development-tooling) though.

File: packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+        if (!this.isValidForwarderSwapQuote(quote)) {
+            throw new Error(AssetSwapQuoterError.InvalidForwarderSwapQuote);
+        }
+        const { ethAmount, feeRecipient, feePercentage: unFormattedFeePercentage } = _.merge(
+            {},
+            constants.DEFAULT_FORWARDER_SWAP_QUOTE_CONSUMER_FEE_OPTS,
+            opts,
+        );
+
+        assert.isValidSwapQuote('quote', quote);
```
Comment (fabioberger) [inline packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts]: What is the difference between this and the above conditional?

File: packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+            assert.isBigNumber('gasPrice', gasPrice);
+        }
+
+        const swapQuoteWithFeeAdded = addAffiliateFeeToSwapQuote(quote, feePercentage);
+
+        const { orders, feeOrders, makerAssetBuyAmount, worstCaseQuoteInfo } = swapQuoteWithFeeAdded;
+
+        const finalTakerAddress = await this._getTakerAddressOrThrowAsync(opts);
+        try {
+            const txHash = await this._contractWrappers.forwarder.marketBuyOrdersWithEthAsync(
```
Comment (fragosti) [inline packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts]: Should this call be using `getSmartContractParamsOrThrow` above to derive the parameters?

File: packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+        }
+    }
+
+    private _getEtherTokenAssetDataOrThrow(): string {
+        return assetDataUtils.getEtherTokenAssetData(this._contractWrappers);
+    }
+
+    private readonly _isValidForwarderSignedOrders = (orders: SignedOrder[]): boolean => {
+        const wethAssetData = this._getEtherTokenAssetDataOrThrow();
+        return _.reduce(orders, (a: boolean, c: SignedOrder): boolean => {
```
Comment (fragosti) [inline packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts]: I feel like a `.filter` + length would be easier to understand here. Or just `every` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every

File: packages/asset-buyer/src/utils/assert.ts [hunk 16]
```diff
@@ -1,25 +1,27 @@
...
-        }
+    isValidSwapQuote(variableName: string, swapQuote: SwapQuote): void {
+        sharedAssert.isHexString(`${variableName}.takerAssetData`, swapQuote.takerAssetData);
+        sharedAssert.isHexString(`${variableName}.makerAssetData`, swapQuote.makerAssetData);
+        sharedAssert.doesConformToSchema(`${variableName}.orders`, swapQuote.orders, schemas.signedOrdersSchema);
+        sharedAssert.doesConformToSchema(`${variableName}.feeOrders`, swapQuote.feeOrders, schemas.signedOrdersSchema);
+        assert.isValidSwapQuoteInfo(`${variableName}.bestCaseQuoteInfo`, swapQuote.bestCaseQuoteInfo);
+        assert.isValidSwapQuoteInfo(`${variableName}.worstCaseQuoteInfo`, swapQuote.worstCaseQuoteInfo);
+        sharedAssert.isBigNumber(`${variableName}.makerAssetBuyAmount`, swapQuote.makerAssetBuyAmount);
+        // TODO(dave4506) Remove once forwarder features are reimplemented
```
Comment (fragosti) [inline packages/asset-buyer/src/utils/assert.ts]: Can this be uncommented?

File: packages/asset-buyer/src/types.ts [hunk 17]
```diff
@@ -25,84 +25,110 @@ export interface OrderProviderResponse {
...
+}
+
+export interface CalldataInformation {
+    calldataHexString: string;
+    to: string;
+    value: BigNumber;
+}
+
+export interface SmartContractParams {
+    params: { [name: string]: any };
```
Comment (fragosti) [inline packages/asset-buyer/src/types.ts]: Might be worth making this generic.
```
export interface SmartContractParams<T> {
    params: T;
    to: string;
    value: BigNumber;
}
```

File: packages/asset-buyer/src/types.ts [hunk 18]
```diff
@@ -25,84 +25,110 @@ export interface OrderProviderResponse {
...
+    to: string;
+    value: BigNumber;
+}
+
+export interface SwapQuoteConsumer {
+    getCalldataOrThrow(quote: SwapQuote, opts: Partial<SwapQuoteGetOutputOpts>): CalldataInformation;
```
Comment (fragosti) [inline packages/asset-buyer/src/types.ts]: Personally I'm not a fan of our `OrThrow` convention (anything can throw at the end of the day). Especially for a public-facing, top-level API, I think it's ok to omit the `OrThrow` but I'd defer to @fabioberger

File: packages/asset-buyer/src/asset_buyer.ts [hunk 19]
```diff
@@ -349,16 +387,19 @@ export class AssetBuyer {
...
+
     /**
-     * Get the assetData that represents the WETH token.
-     * Will throw if WETH does not exist for the current network.
+     *
+     * get the key for _orderEntryMap for maker + taker asset pair
      */
-    private _getEtherTokenAssetDataOrThrow(): string {
-        return assetDataUtils.getEtherTokenAssetData(this._contractWrappers);
+    // tslint:disable-next-line: prefer-function-over-method
```
[the remark this answers is not in the document]
Comment (dave4506) [reply packages/asset-buyer/src/asset_buyer.ts]: static would make the function public?

**Comment by dave4506** [comment]
Per comment on confidence, I have refactored the previous tests, I am fairly confident in that the `SwapQuoter` works as intended. The `ForwarderSwapQuoteConsumer` has been only manually tested (I fed in SwapQuotes and looked at the output of the consumer). Because of the logical simplicity of `ForwarderSwapQuoteConsumer`, I am comfortable with the lack of test coverage rn with the consumer, (will be adding it later).

File: packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts (hunk 14 above)
Comment (fragosti) [reply packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts]: @dave4506 not sure I understand.

File: packages/asset-buyer/test/forwarder_swap_quote_consumer_test.ts [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+                const invalidSignedOrders = getSignedOrdersWithNoFees(
+                    makerAssetData,
+                    takerAssetData,
+                    makerAddress,
+                    takerAddress,
+                    FILLABLE_AMOUNTS,
+                );
+                const invalidSwapQuote = getFullyFillableSwapQuoteWithNoFees(makerAssetData, takerAssetData, invalidSignedOrders);
+                const swapQuoteConsumer = new ForwarderSwapQuoteConsumer(provider, {});
+                // TODO(dave4506) finish up testing/coverage
```
Comment (fragosti) [inline packages/asset-buyer/test/forwarder_swap_quote_consumer_test.ts]: Is this intentional?

Comment (fragosti) [reply packages/asset-buyer/test/forwarder_swap_quote_consumer_test.ts]: if it helps you need to pass in a function that throws, not an expression that throws.

File: packages/asset-buyer/src/utils/affiliate_fee_utils.ts [hunk 21]
```diff
@@ -0,0 +1,29 @@
+import { BigNumber } from '@0x/utils';
+import * as _ from 'lodash';
+
+import { SwapQuote, SwapQuoteInfo } from '../types';
+
+export const affiliateFeeUtils = {
+    getSwapQuoteWithAffiliateFee(quote: SwapQuote, feePercentage: number): SwapQuote {
```
Comment (fragosti) [inline packages/asset-buyer/src/utils/affiliate_fee_utils.ts]: I still feel like these two `SwapQuotes` should be different types for clarity. Ie. `SwapQuote` vs. `SwapQuoteWithAffiliateFee`

File: packages/asset-buyer/test/forwarder_swap_quote_consumer_test.ts [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+                    wethAssetData,
+                    makerAddress,
+                    takerAddress,
+                    FILLABLE_AMOUNTS,
+                );
+                const swapQuote = getFullyFillableSwapQuoteWithNoFees(makerAssetData, takerAssetData, signedOrders);
+                const swapQuoteConsumer = new ForwarderSwapQuoteConsumer(provider, { networkId });
+                const callDataInfo = await swapQuoteConsumer.getCalldataOrThrowAsync(swapQuote, {});
+                // console.log(callDataInfo);
+                // TODO(dave4506): Add elaborate testing
```
Comment (fragosti) [inline packages/asset-buyer/test/forwarder_swap_quote_consumer_test.ts]: This test isn't finished. I don't think it should run / be in the code if it isn't finished.

File: packages/asset-buyer/src/asset_buyer.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (fragosti) [reply packages/asset-buyer/src/asset_buyer.ts]: static functions can be private? But this tslint comment is fine.

File: packages/asset-buyer/test/forwarder_swap_quote_consumer_test.ts (hunk 20 above)
Comment (dave4506) [reply packages/asset-buyer/test/forwarder_swap_quote_consumer_test.ts]: This is corrected in another PR: https://github.com/0xProject/0x-monorepo/pull/1896

File: packages/asset-buyer/test/forwarder_swap_quote_consumer_test.ts (hunk 22 above)
Comment (dave4506) [reply packages/asset-buyer/test/forwarder_swap_quote_consumer_test.ts]: Will be addressed later: https://github.com/0xProject/0x-monorepo/pull/1896

File: packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts (hunk 14 above)
Comment (dave4506) [reply packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts]: That was a crappy explanation: the contract wrappers parameters do not map 1 to 1 to smart contract params. They do some additional logic to construct the smart contract params, redundant to do that work again before we call execute.

**Pull request merged**